controllers/gce/utils/utils_test.go

```diff
@@ -0,0 +1,482 @@
+		rk := namer.Truncate(fooTest.k)
+		if rk != fooTest.ek {
+			t.Errorf("returned %v but expected %v", rk, fooTest.ek)
+		}
+	}
+}
+
+func TestDecorateName(t *testing.T) {
+	nameLenLimitValue := ""
+	for i := 0; i < nameLenLimit-6; i++ {
+		nameLenLimitValue = fmt.Sprintf("%s%s", nameLenLimitValue, "t")
+	}
+
+	fooTests := []struct {
+		cn  string
+		n   string
+		edn string
+	}{
+		{"", "", ""},
+		{"", "fooName", "fooName"},
+		{"fooCN", "fooName", "fooName--fooCN"},
+		{"fooCN", nameLenLimitValue, nameLenLimitValue + "--fooC0"},
+	}
+
+	for _, fooTest := range fooTests {
+		namer := NewNamer(fooTest.cn, "fooFN")
+		rdn := namer.decorateName(fooTest.n)
+		if rdn != fooTest.edn {
+			t.Errorf("returned %v but expected %v", rdn, fooTest.edn)
+		}
+	}
+}
+
+func TestParseName(t *testing.T) {
+	fooTests := []struct {
+		n   string
+		ecn string
+		ers string
+	}{
+		{"", "", ""},
+		{"foo", "", ""},
+		{"foo01--", "", ""},
+		{"foo01--foo02", "foo02", ""},
+		{"foo01-foo02--foo03", "foo03", "foo02"},
+	}
+
+	namer := NewNamer("fooCN", "fooFN")
+	for _, fooTest := range fooTests {
+		nc := namer.ParseName(fooTest.n)
+		if nc == nil {
+			t.Errorf("returned nil")
+			continue
+		}
+
+		if nc.ClusterName != fooTest.ecn {
+			t.Errorf("returned %v but expected %v", nc.ClusterName, fooTest.ecn)
+			continue
+		}
+
+		if nc.Resource != fooTest.ers {
+			t.Errorf("returned %v but expected %v", nc.Resource, fooTest.ers)
+			continue
+		}
+	}
+}
+
+func TestNameBelongsToCluster(t *testing.T) {
+	fooTests := []struct {
+		n  string
+		cn string
+		ev bool
+	}{
+		{"", "", false},
+		{"fooName", "", false},
+		{"k8s-fooName", "fooCN", false},
+		{"k8s-fooName", "", true},
+		{"k8s-foo01--foo02--foo03", "foo", false},
+		{"k8s-foo01--foo02", "foo02", true},
+		{"k8s-foo01--foo02", "", false},
+		{"k8s-foo01--foo02", "foo", false},
+	}
+
+	namer := NewNamer("fooCN", "fooFN")
+	for _, fooTest := range fooTests {
+		namer.SetClusterName(fooTest.cn)
+		rv := namer.NameBelongsToCluster(fooTest.n)
+		if rv != fooTest.ev {
+			t.Errorf("returned %v but expected %v", rv, fooTest.ev)
+		}
+	}
+}
+
+func TestBeName(t *testing.T) {
+	fooTests := []struct {
+		cn string
+		p  int64
+		en string
+	}{
+		{"", 80, "k8s-be-80"},
+		{"fooCN", 80, "k8s-be-80--fooCN"},
+	}
+
+	for _, fooTest := range fooTests {
+		namer := NewNamer(fooTest.cn, "fooFN")
+		rn := namer.BeName(fooTest.p)
+		if rn != fooTest.en {
+			t.Errorf("returned %v but expected %v", rn, fooTest.en)
+		}
+	}
+}
+
+func TestBePort(t *testing.T) {
+	namer := NewNamer("fooCN", "fooFN")
+
+	fooTests := []struct {
+		bn string
+		ie bool
+		bp string
+	}{
+		{"", true, ""},
+		{"fooBP", true, ""},
+		{"k8s-be-1022 k8s-be-033", false, "1022"},
+		{"k8s-be-022 k8s-be-033", false, "022"},
+	}
+
+	for _, fooTest := range fooTests {
+		rbp, err := namer.BePort(fooTest.bn)
+		if fooTest.ie {
+			if err == nil {
+				t.Errorf("expected error")
+			}
+			continue
+		}
+
+		if err != nil {
+			t.Errorf("unexpected error")
```

Comment:
If error really exist, the next test case still could be and should be executed, so i I think it's just right now, do you think so? thank you!